### ISSUE TITLE
TensorRT outputs -1 when running modulus on certain integer values

### ISSUE THREAD

**Issue by wxuwaabi** [NVIDIA/TensorRT#4702]
## Description

TensorRT produces incorrect results for the modulus operation (`x % 32`) on int32 tensors for certain input values. The engine returns `-1` instead of the correct result `31` for inputs like `18542495` and `2147483647`, while working correctly for other values like `1800000000`.

This appears to be a correctness bug in how TensorRT handles the modulus operation on signed int32 values. The issue also persists when using int64 rather than int32.

**Error Example**:
```
✗ Test failed for input 18542495
  TensorRT output: -1
  PyTorch output: 31
```

## Environment

**TensorRT Version**: 10.7.0.post1

**NVIDIA GPU**: NVIDIA A10G

**NVIDIA Driver Version**: 570.172.08

**CUDA Version**: 12.6

**CUDNN Version**: 90501

**Operating System**: Ubuntu 22.04.4 LTS

**Python Version (if applicable)**: 3.10.7

**Tensorflow Version (if applicable)**: N/A

**PyTorch Version (if applicable)**: 2.6.0.post3

**Baremetal or Container (if so, version)**: Baremetal

## Relevant Files

**Model link**: Minimal model included in reproduction script below (simple modulus operation)

### Full Reproducible Script

```python
import tensorrt as trt
import torch


ONNX_FILE = "modulus.onnx"

example_tensor = torch.tensor([1], dtype=torch.int32, device="cuda")


class ExampleModulusModule(torch.nn.Module):
    def forward(self, x):
        return x % 32


module = ExampleModulusModule()
with torch.no_grad():
    torch.onnx.export(
        module,
        (example_tensor,),
        ONNX_FILE,
        opset_version=18,
        input_names=["input"],
        output_names=["output"],
        dynamo=True
    )


# Create TensorRT logger and builder
trt_logger = trt.Logger(trt.Logger.INFO)
trt_builder = trt.Builder(trt_logger)
trt_network = trt_builder.create_network()
trt_parser = trt.OnnxParser(trt_network, trt_logger)
trt_config = trt_builder.create_builder_config()

# Configure TensorRT settings
trt_config.profiling_verbosity = trt.ProfilingVerbosity.DETAILED
trt_config.hardware_compatibility_level = trt.HardwareCompatibilityLevel.AMPERE_PLUS

# Parse the ONNX file
with open(ONNX_FILE, "rb") as f:
    if not trt_parser.parse(f.read()):
        for i in range(trt_parser.num_errors):
            print(trt_parser.get_error(i))
        raise RuntimeError("Failed to parse the ONNX file")

# Build the serialized network
print("Building TensorRT engine...")
engine_bytes = trt_builder.build_serialized_network(trt_network, trt_config)

# Deserialize the engine
runtime = trt.Runtime(trt_logger)
engine = runtime.deserialize_cuda_engine(engine_bytes)

# Create execution context
context = engine.create_execution_context()

# Get tensor names and shapes
input_name = "input"
output_name = "output"

input_shape = engine.get_tensor_shape(input_name)
output_shape = engine.get_tensor_shape(output_name)

print(f"Input shape: {input_shape}")
print(f"Output shape: {output_shape}")

# Run test cases
print("\n" + "=" * 80)
print("Testing TensorRT Engine")
print("=" * 80)

for test_case in [18542495, 1800000000, 2147483647]:
    input_tensor = torch.tensor([test_case], dtype=torch.int32, device="cuda").contiguous()
    output_tensor = torch.empty(tuple(output_shape), dtype=torch.int32, device="cuda").contiguous()

    # Set tensor addresses
    context.set_tensor_address(input_name, input_tensor.data_ptr())
    context.set_tensor_address(output_name, output_tensor.data_ptr())

    # Execute inference
    context.execute_async_v3(torch.cuda.current_stream().cuda_stream)
    torch.cuda.synchronize()

    # Compare with PyTorch result
    expected_out = module(input_tensor)

    try:
        torch.testing.assert_close(output_tensor, expected_out)
        print(f"✓ Test passed for input {test_case}: TRT={output_tensor.item()}, PyTorch={expected_out.item()}")
    except AssertionError as e:
        print(f"✗ Test failed for input {test_case}")
        print(f"  TensorRT output: {output_tensor.item()}")
        print(f"  PyTorch output: {expected_out.item()}")

print("=" * 80)
```

**Commands or scripts**:

```bash
python reproduce_modulus_bug.py
```

**Output**:
```
================================================================================
Testing TensorRT Engine
================================================================================
✗ Test failed for input 18542495
  TensorRT output: -1
  PyTorch output: 31
✓ Test passed for input 1800000000: TRT=0, PyTorch=0
✗ Test failed for input 2147483647
  TensorRT output: -1
  PyTorch output: 31
```

**Results Summary**:

| Input       | Expected (PyTorch) | TensorRT Result | Status |
|-------------|-------------------|-----------------|--------|
| 18542495    | 31                | -1              | ❌ FAIL |
| 1800000000  | 0                 | 0               | ✅ PASS |
| 2147483647  | 31                | -1              | ❌ FAIL |

**Have you tried [the latest release](https://developer.nvidia.com/tensorrt)?**:

No

**Attach the captured .json and .bin files from [TensorRT's API Capture tool](https://docs.nvidia.com/deeplearning/tensorrt/latest/inference-library/capture-replay.html) if you're on an x86_64 Unix system**

N/A

**Can this model run on other frameworks?** For example run ONNX model with ONNXRuntime (`polygraphy run <model.onnx> --onnxrt`):

It should be possible, this is a trivial model

**Comment by CodersAcademy006** [comment]
Hey @wxuwaabi Thanks for the detailed repro, this is a nice minimal example.

One thing that stands out is that the failing cases are where the divisor is a **power of two** (`32`) and the input values are fairly large (`18542495`, `2147483647`). In many runtimes `% 2ⁿ` sometimes gets optimized internally into a bitmask operation instead of a full integer remainder.

If that optimization path is applied with the wrong signed semantics, it can lead to results like `-1` instead of `31`. PyTorch follows the Euclidean remainder definition for `%`, while some lower-level implementations treat it closer to a truncating remainder or apply masking assuming unsigned values.

A quick sanity check would be to try the same model with a **non power-of-two divisor** (for example `% 31` or `% 33`). If those work correctly, that would strongly suggest the optimized `% 2ⁿ` path is the one causing the mismatch.

If that’s indeed what’s happening, the fix likely needs to happen in the step where the ONNX `Mod` op gets lowered to TensorRT kernels. I can take a look at that path as well if it helps.

**Comment by wxuwaabi** [comment]
Hey @CodersAcademy006 , thanks for taking a look! The issue seems to persist with arbitrary divisors, I modified the script to run on chunks and ran it for DIVISOR=3 and 31. It does seem like inputs greater than `16000000` trigger this issue, all numbers below `16000000` seem fine:

```
import tensorrt as trt
import torch


ONNX_FILE = "modulus.onnx"
CHUNK_SIZE = 1_000_000
INT32_MAX = 2_147_483_647
DIVISOR = 3

example_tensor = torch.zeros(CHUNK_SIZE, dtype=torch.int32, device="cuda")


class ExampleModulusModule(torch.nn.Module):
    def forward(self, x):
        return x % DIVISOR


module = ExampleModulusModule()
with torch.no_grad():
    torch.onnx.export(
        module,
        (example_tensor,),
        ONNX_FILE,
        opset_version=18,
        input_names=["input"],
        output_names=["output"],
        dynamo=True
    )


# Create TensorRT logger and builder
trt_logger = trt.Logger(trt.Logger.INFO)
trt_builder = trt.Builder(trt_logger)
trt_network = trt_builder.create_network()
trt_parser = trt.OnnxParser(trt_network, trt_logger)
trt_config = trt_builder.create_builder_config()

# Configure TensorRT settings
trt_config.profiling_verbosity = trt.ProfilingVerbosity.DETAILED
trt_config.hardware_compatibility_level = trt.HardwareCompatibilityLevel.AMPERE_PLUS

# Parse the ONNX file
with open(ONNX_FILE, "rb") as f:
    if not trt_parser.parse(f.read()):
        for i in range(trt_parser.num_errors):
            print(trt_parser.get_error(i))
        raise RuntimeError("Failed to parse the ONNX file")

# Build the serialized network
print("Building TensorRT engine...")
engine_bytes = trt_builder.build_serialized_network(trt_network, trt_config)

# Deserialize the engine
runtime = trt.Runtime(trt_logger)
engine = runtime.deserialize_cuda_engine(engine_bytes)

# Create execution context
context = engine.create_execution_context()

# Get tensor names and shapes
input_name = "input"
output_name = "output"

input_shape = engine.get_tensor_shape(input_name)
output_shape = engine.get_tensor_shape(output_name)

print(f"Input shape: {input_shape}")
print(f"Output shape: {output_shape}")

# Run test cases
print("\n" + "=" * 80)
print("Testing TensorRT Engine (chunks of 1,000,000 from 0 to INT32_MAX)")
print("=" * 80)

failed_total = 0
chunk_count = 0

for start in range(0, INT32_MAX + 1, CHUNK_SIZE):
    end = min(start + CHUNK_SIZE, INT32_MAX + 1)
    actual_size = end - start

    chunk = torch.arange(start, end, dtype=torch.int32, device="cuda")
    if actual_size < CHUNK_SIZE:
        chunk = torch.nn.functional.pad(chunk, (0, CHUNK_SIZE - actual_size))
    input_tensor = chunk.contiguous()
    output_tensor = torch.empty(CHUNK_SIZE, dtype=torch.int32, device="cuda").contiguous()

    context.set_tensor_address(input_name, input_tensor.data_ptr())
    context.set_tensor_address(output_name, output_tensor.data_ptr())

    context.execute_async_v3(torch.cuda.current_stream().cuda_stream)
    torch.cuda.synchronize()

    expected = chunk[:actual_size] % DIVISOR
    actual = output_tensor[:actual_size]

    mismatches = int((actual != expected).sum().item())
    if mismatches > 0:
        failed_total += mismatches
        first_bad = int((actual != expected).nonzero(as_tuple=True)[0][0].item())
        print(f"✗ Chunk [{start}, {end}): {mismatches} mismatches, e.g. input={start + first_bad}, TRT={actual[first_bad].item()}, expected={expected[first_bad].item()}")
    else:
        print(f"✓ Chunk [{start}, {end}): all {actual_size} values correct")

    chunk_count += 1

print("=" * 80)
if failed_total == 0:
    print(f"All {chunk_count} chunks passed.")
else:
    print(f"{failed_total} total mismatches across {chunk_count} chunks.")
print("=" * 80)
```
